### PR TITLE
 support register new terror

### PR DESCRIPTION
**Pull request by lysu**
### What is changed and how it works?

make terror support register new error code and error msg and prepare for move error code/msg out of parser

### Check List 

Tests 

 - Unit test
 - Integration test

Code changes

 - impl

Side effects

 - n/a

Related changes

 - n/a

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/parser/734)

### CODE REVIEW

File: terror/terror.go [hunk 1]
```diff
@@ -144,6 +128,15 @@ func (ec ErrClass) NotEqualClass(err error) bool {
 // New creates an *Error with an error code and an error message.
 // Usually used to create base *Error.
 func (ec ErrClass) New(code ErrCode, message string) *Error {
+	if ErrClassToMySQLCodes == nil {
+		ErrClassToMySQLCodes = make(map[ErrClass]map[ErrCode]struct{})
```
Comment (kennytm) [inline terror/terror.go]: Any chance this function is called simultaneously from two different goroutines?

Comment (lysu) [reply terror/terror.go]: no this, function only can be called in `init()` or `global variable initializer`,  I will add some comments for this :D

Comment (tiancaiamao) [reply terror/terror.go]: But there is no need for this lazy initialization @lysu

File: terror/terror.go [hunk 2]
```diff
@@ -291,21 +284,20 @@ func (e *Error) getMySQLErrorCode() uint16 {
...
 
 var (
-	// ErrClassToMySQLCodes is the map of ErrClass to code-map.
-	ErrClassToMySQLCodes map[ErrClass]map[ErrCode]uint16
+	// ErrClassToMySQLCodes is the map of ErrClass to code-set.
+	ErrClassToMySQLCodes map[ErrClass]map[ErrCode]struct{}
 )
 
 func init() {
-	ErrClassToMySQLCodes = make(map[ErrClass]map[ErrCode]uint16)
```
Comment (tiancaiamao) [inline terror/terror.go]: We can keep this way to initialize `ErrClassToMySQLCodes`.

**Comment by lysu** [comment]
==  it has some race problem need fix

**Comment by lysu** [comment]
@jackysp @zimulala  hi, I have do some change, add a new `Synthesize`, because I found TiDB will create terror base on TiKV's error code in cop select_result.go, we can not register TiKV error code directly in current logic(maybe it's not right..),  so add a new method just create terror but didn't register code - -

**Pull request merged**